text/1025-branches.rst

```diff
@@ -0,0 +1,181 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2024-01-11
+    Authors: Victor Petrovykh <victor@edgedb.com>
+
+
+===========================
+RFC 1025: Database Branches
+===========================
+
+Abstract
+========
+
+We propose to reframe the current concept of ``database`` in EdgeDB as
+``branch``. This avoids some of the confusion in current terminology and
+provides a clearer guideline for using EdgeDB in development environment.
+Functionally a ``branch`` will be the same as the current ``database``, but
+with some additional options and tooling.
+
+
+Motivation
+==========
+
+The current nomenclature of ``database`` and the corresponding commands are
+confusing to many users. A running EdgeDB server represents an *instance*
+while each instance can have multiple *databases* in it. Since the databases
+all run by the same server they all have to use the same underlying EdgeDB
+version. These qualities make them impractical for usage across multiple
+projects as the projects would have to share resources and perform EdgeDB
+upgrades together. We specifically recommend to have a dedicated instance for
+each project. This leaves databases as a feature without a clear usage.
+
+Instead we want to enhance the developer experience while using multiple
+source control branches. These multiple databases could map well onto
+corresponding source branches, representing slightly different version of the
+data and schema. Thus there is a good reason to rename the concept of
+*database* to *branch* and provide additional branch-handling tools. With this
+change it is also now no longer necessary to have a semantic distinction
+between the terms *instance* and *database* which may be natural for some
+users to think of as interchangeable.
+
+
+Branch DDL Commands
+===================
+
+There are several new DDL commands introduced by this proposal:
+
+1) ``create empty branch <newbranch>``
+
+   The most basic command creates a new branch with an empty schema
+   (exactly what ``create database`` command currently does).
+
+2) ``create branch <newbranch> [ from <oldbranch> ]``
+
+   This command creates a new branch and copies the schema of an existing
+   branch to it. Only the schema is copied, the data is still empty and needs
+   to be populated separately. This parallels source control branches that
+   diverge in terms of the schema. It is recommended to setup a new database
+   branch whenever working on a code branch that alters the schema in some
+   way. This will help reduce friction when switching between code branches by
+   having database states that correspond to the schema and avoiding
+   accidental mismatches in migrations.
+
+   If the ``from <oldbranch>`` part is omitted, the new branch is created
+   using the branch from the current connection.
```

Comment:
I think that @1st1 specifically didn't want this, and wanted to force being more explicit